### ISSUE TITLE
From WebUI to the history server

### ISSUE THREAD

**Issue by superbobry** [jcrist/skein#77]
Now that Skein has the in-progress UI, I think it is time to go one step further and implement a history server. The goal of the history server is to display a fixed number of recently finished Skein applications and serve a snapshot of the in-progress UI for each application on request.

In the simplest version we could make the AM (or the Daemon, not sure which is more appropriate) persist the `UIModel` on a preconfigured directory on HDFS once the application has finished. The history server would monitor this directory and list its contents in a table (see Spark history for an example). Then for each specific app the history server could load the `UIModel` and reuse `WebUI` to render it. 

Wdyt?

**Comment by jcrist** [comment]
Apologies for the delayed response here. My personal thought is that this would be a lot of work, and wouldn't be that useful. The bare web UI contains:

- Status of containers for each service
- The Key-Value store

The first is will just report all containers as completed, failed, and the latter is really more meant for debugging. Much of this information is available via YARN resource-manager web ui, or the application logs (which continue to be available via the YARN web-ui).

The Spark history server is more useful because it reports more application-specific information. This is what #72 is for, but we wouldn't want to save copies of the proxied pages (wouldn't want to crawl all links in the page). Instead, I think persisting application logs or implementing a "history server" should be left for user applications (e.g. dask-yarn may later implement it's own history server), as the Skein web-ui will never be application-specific enough to have a generic history server be useful.

**Comment by superbobry** [comment]
I agree with your general point. The motivation for creating this issue comes from the fact that the YARN UI does not list containers used by the application, neither while the application is running, nor after. This makes it difficult to debug failures: one has to look up the logs via CLI and then manually inspect the output for errors. Imagine doing so for 100 containers which is easily possible in the case of TensorFlow.

**Comment by jcrist** [comment]
That's a fair point, but I'm not 100% convinced that a history server is the best way to resolve this. A few assorted thoughts:

- Continuing to serve the same web-ui after completion is nice from a uniformity point, but has the following complications:

  - User-registered pages will no longer be active. This is fine, we can just choose not to display them, but it's not 100% the same experience.
  - The container logs have moved. We'd need to also copy the logs over, or serve them from the aggregated location (if configured). I'm not sure how to find the aggregated log location from the yarn configuration, but it's definitely doable.

- In general all you'd want in a debugging scenario is access to the logs and the container final states. This information is present in the web ui, but could be in a different solution as well.

For now, I I'd like to think about this problem for a bit. I see a few options:

- Full blown history server that may be optionally running 24/7. The application web ui would be served from the server at all points, meaning that the `history` link in the YARN webui would always work (provided the history logs were still present). This would match other tools (e.g. Spark), and makes nice use of all of YARN's facilities. Downside is this is a lot more work, would require cluster admin buy-in in many environments, as well as a larger configuration setup. I'm currently -0.5 on this. So far Skein has been designed to not require any admin level permissions or buy-in for any component.

- User-run history server. User could optionally persist some information on hdfs, and serve it later with something like `skein history serve <PATH>`. They'd have to tunnel-out the server from edge node to see it, but this may work if using it is unlikely. History links on the YARN webui would continue to not work. This seems easier, but also less useful. I'm unlikely to work on this method.

- Store information on container exit states *in* the logs in a special format. YARN collects logs from a directory, we can have the application master write information to a separate file in that directory that can be parsed later. Then an external tool could be written to parse the logs and browse them in a useful way. No need for us to persist information externally, all of it is written to YARN's normal logs. This also opens up use of a cli-based browser in addition to a web-based one. I'm currently leaning towards this option. This could also later be changed into the first option if needed.

I don't currently plan to work on this in the near future, but would be interested in hearing your thoughts.
